### PR TITLE
[WIP] customizable export UI for DistortableCollection

### LINKED ISSUE #478: Allow over-riding startExport on Collections, to enable different remote exporters to be attached
Related to #474 and #471 cc @sashadev-sky ! 🎉 

I'm starting to work on this!

### PR DESCRIPTION
**Pull request by jywarren**
Fixes #478 

cc @sashadev-sky @SidharthBansal this is very rough, but I'm making settable options params for all critical export UI and request/response handling; I'm seeing some issue with variable scope, however, so this needs a lot of debugging. I want to demo how setting a custom Exporter and Exporter UI work in the new examples/export.html file, too!

If we can get GCI folks engaged in this, that'd be amazing!

### CODE REVIEW

File: src/edit/DistortableCollection.Edit.js [hunk 1]
```diff
@@ -8,9 +8,88 @@ L.DistortableCollection.Edit = L.Handler.extend({
 
   initialize: function(group, options) {
     this._group = group;
+
+// OK let's just refactor this to not use the closure scope, but to pass options into each function.
```
Comment (jywarren) [inline src/edit/DistortableCollection.Edit.js]: This is my next step to try to resolve the bug i was seeing.

File: src/edit/actions/ExportAction.js [hunk 2]
```diff
@@ -29,7 +29,18 @@ L.ExportAction = L.EditAction.extend({
       L.IconUtil.toggleXlink(this._link, 'get_app', 'spinner');
       L.IconUtil.toggleTitle(this._link, 'Export Images', 'Loading...');
       L.IconUtil.addClassToSvg(this._link, 'loader');
-      edit.startExport();
+console.log(this, edit.options)
+// explicitly send each option:
+      edit.startExport({
+        collection: edit.options.collection,
```
Comment (jywarren) [inline src/edit/actions/ExportAction.js]: Here we can see all the settable options being passed into the export start command; there are a lot, allowing for some really detailed customization of both the Export process management and the UI.

File: src/edit/DistortableCollection.Edit.js [hunk 3]
```diff
@@ -8,9 +8,94 @@ L.DistortableCollection.Edit = L.Handler.extend({
...
+// OK let's just refactor this to not use the closure scope, but to pass options into each function. 
+// this is important anyways so you can override them from the HTML page and access private values
+
+    this.startExport = options.startExport || function startExport(opts) {
+      return new Promise(function(resolve) {
+        opts = opts || {};
+
+    // this is undefined at runtime but gets filled in later... is this an async issue?
+    // ok, wrapped this in a promise
+    console.log('options in startExport', opts, opts.exportUrl);
```
Comment (jywarren) [inline src/edit/DistortableCollection.Edit.js]: So, `opts.exportUrl` returns `undefined` but `opts` shows correct values, so we're seeing a timing or scope error here; this was the case also before adding the Promise code on line 16, which I had hoped would solve it. Could there be a scope collision with the `opts` parameter somehow?

File: examples/export.html [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            "cm_per_pixel":35.8578,
+            "src":"https://s3.amazonaws.com/grassrootsmapping/warpables/320983/test.png"}
+        ];
+
+        // initialize the collection:
+        imgGroup = L.distortableCollection({
+          collection: json,
+// this needs to be a function run at the moment of pressing, not when initializing the collection
+//          scale: prompt("Choose a scale or use the default (cm per pixel):", 100),
+          exportUrl: 'http://34.74.118.242/api/v2/export/', // used to
```
Comment (jywarren) [inline examples/export.html]: Here I'm showing one reason WHY we want to be able to customize; we have alternative export servers with slightly different requirements, which we'd like to be able to connect to the Exporter in a simple standardized way.

**Comment by jywarren** [comment]
@sashadev-sky any ideas on what I'm doing wrong here? My brain is a bit fried and I have a suspicion it's something pretty simple.

**Comment by SidharthBansal** [comment]
@sashadev-sky can you please upload this task on the GCI dashboard by summarizing it. I am afraid I will not be able to properly summarise it as I don't have experience of LDI

@jywarren should it be a hard task for GCI?

File: src/edit/DistortableCollection.Edit.js (hunk 3 above)
Comment (jywarren) [reply src/edit/DistortableCollection.Edit.js]: Solved this in latest code i'm pushing!

**Comment by jywarren** [comment]
Got this working! Now we need to handle the response!

**Comment by jywarren** [comment]
OK, it's basically working now! We overrode the actual images using a JSON collection, which we don't have to do, but demonstrates how we can filter the actual images submitted. We could further refine this but the key parts we need to wrap up a first implementation and merge it are:

* [ ] stopping the spinner, which i think is something we can do as of recently?

Anything else?

Then we should:

* [ ] write more docs
* [ ] try to auto-calculate appropriate resolution using `_getAvgCmPerPixel()`
* [ ] connect up with other export output formats in https://github.com/publiclab/image-sequencer-app/issues/36 like data-urls, or Google Storage-backed exports?
* [ ] load test it with larger image sets in coordination with @tech4GT  and https://github.com/publiclab/image-sequencer-app/issues/19 !!! 🎉

File: src/edit/DistortableCollection.Edit.js [hunk 5]
```diff
@@ -29,7 +104,7 @@ L.DistortableCollection.Edit = L.Handler.extend({
       singleclickon: this._singleClickListeners,
       singleclickoff: this._resetClickListeners,
       singleclick: this._singleClick,
-      boxcollectend: this._addCollections,
+      boxzoomend: this._addCollections,
```
Comment (jywarren) [inline src/edit/DistortableCollection.Edit.js]: Oops, we shouldn't be overwriting this recent improvement. Reverting!

**Comment by jywarren** [comment]
I passed the `resolve()` into options so it can be resolved from the customized export functions 😄 🎉

**Comment by jywarren** [comment]
So perhaps the follow-ups could become GCI issues now?

**Comment by jywarren** [comment]
Hmm, I tried resolving the linting errors with reordering but may have made some mistakes. They are pretty minor though, once we resolve them we can merge this.

**Comment by jywarren** [comment]
All good here! @sashadev-sky can you give a final review after the holidays? 😄

**Comment by jywarren** [comment]
We could also write a test against this by overriding the export functions and using a mock function to callback or something... interesting!

**Comment by jywarren** [comment]
Let's circle back to https://github.com/publiclab/mapknitter/issues/1128 once we're done here, too!

**Comment by jywarren** [comment]
https://github.com/publiclab/Leaflet.DistortableImage/pull/501 adds tests on top of this but I wasn't able to get them to work... we can circle back later to finalize that!

**Comment by sashadev-sky** [comment]
@jywarren there are a bunch of GCI PRs open right now for me to review! Let me know where this stands priority wise, assuming the others get precedence for now

**Comment by SidharthBansal** [comment]
Hi, Sasha we need to review all the prs of GCI merged before GCI deadline.

**Comment by sashadev-sky** [comment]
Yes that I know for sure don’t worry! That’s not a prob

**Comment by jywarren** [comment]
This is actually really high priority if you're at all able to take a look. We are pretty urgently trying to get MapKnitter off of it's internal exporter to reduce the cloud server charges we incur from it. I'd appreciate if you could take a look! Thank you!!

**Comment by jywarren** [comment]
I think this is safe to merge and we can add the tests next.

**Comment by jywarren** [comment]
I'm going to go ahead and merge this, actually! It's got some key fixes needed in https://github.com/publiclab/mapknitter/issues/1191 -- thank you!!!

**Comment by jywarren** [comment]
I'd still love a review however 😄

**Comment by SidharthBansal** [comment]
Awesome folks.

**Pull request merged**